crates/oxc_semantic/src/builder.rs

```diff
@@ -306,6 +306,14 @@ impl<'a> SemanticBuilder<'a> {
         self.errors.borrow_mut().push(error);
     }
 
+    pub(crate) fn in_declare_scope(&self) -> bool {
+        self.source_type.is_typescript_definition()
+            || self
```

Comment:
This is expensive, try use the ambient flag - move the check logic to the parser if semantic don't have this context.

Reply:
i don't think we've enough info in the parser, but i can try?

what do you think about introducing a self.flags for storing this?
that shouldn't incur the same perf cost